tools/tm-signer-harness/internal/test_harness.go

```diff
@@ -0,0 +1,391 @@
+// when exiting the main program.
+type TestHarnessError struct {
+	Code int    // The exit code to return
+	Err  error  // The original error
+	Info string // Any additional information
+}
+
+var _ error = (*TestHarnessError)(nil)
+
+// TestHarness allows for testing of a remote signer to ensure compatibility
+// with this version of Tendermint.
+type TestHarness struct {
+	addr             string
+	sc               *privval.SocketVal
+	fpv              *privval.FilePV
+	chainID          string
+	acceptRetries    int
+	logger           log.Logger
+	exitWhenComplete bool
+	exitCode         int
+}
+
+// TestHarnessConfig provides configuration to set up a remote signer test
+// harness.
+type TestHarnessConfig struct {
+	BindAddr string
+
+	KeyFile     string
+	StateFile   string
+	GenesisFile string
+
+	AcceptDeadline time.Duration
+	ConnDeadline   time.Duration
+	AcceptRetries  int
+
+	SecretConnKey ed25519.PrivKeyEd25519
+
+	ExitWhenComplete bool // Whether or not to call os.Exit when the harness has completed.
+}
+
+// timeoutError can be used to check if an error returned from the netp package
+// was due to a timeout.
+type timeoutError interface {
+	Timeout() bool
+}
+
+// NewTestHarness will load Tendermint data from the given files (including
+// validator public/private keypairs and chain details) and create a new
+// harness.
+func NewTestHarness(logger log.Logger, cfg TestHarnessConfig) (*TestHarness, error) {
+	keyFile := ExpandPath(cfg.KeyFile)
+	stateFile := ExpandPath(cfg.StateFile)
+	logger.Info("Loading private validator configuration", "keyFile", keyFile, "stateFile", stateFile)
+	// NOTE: LoadFilePV ultimately calls os.Exit on failure. No error will be
+	// returned if this call fails.
+	fpv := privval.LoadFilePV(keyFile, stateFile)
+
+	genesisFile := ExpandPath(cfg.GenesisFile)
+	logger.Info("Loading chain ID from genesis file", "genesisFile", genesisFile)
+	st, err := state.MakeGenesisDocFromFile(genesisFile)
+	if err != nil {
+		return nil, newTestHarnessError(ErrFailedToLoadGenesisFile, err, genesisFile)
+	}
+	logger.Info("Loaded genesis file", "chainID", st.ChainID)
+
+	sc, err := newTestHarnessSocketVal(logger, cfg)
+	if err != nil {
+		return nil, newTestHarnessError(ErrFailedToCreateListener, err, "")
+	}
+
+	return &TestHarness{
+		addr:             cfg.BindAddr,
+		sc:               sc,
+		fpv:              fpv,
+		chainID:          st.ChainID,
+		acceptRetries:    cfg.AcceptRetries,
+		logger:           logger,
+		exitWhenComplete: cfg.ExitWhenComplete,
+		exitCode:         0,
+	}, nil
+}
+
+// Run will execute the tests associated with this test harness. The intention
+// here is to call this from one's `main` function, as the way it succeeds or
+// fails at present is to call os.Exit() with an exit code related to the error
+// that caused the tests to fail, or exit code 0 on success.
+func (th *TestHarness) Run() {
+	th.logger.Info("Starting test harness")
+	donec := make(chan struct{})
+	go func() {
```

Comment:
why spawn a goroutine if you block in the end?

Reply:
Good catch! This was a remnant of an older approach that I did away with. Refactored now to remove the goroutine.